### PR TITLE
Port functorch decomps over and fix some tests

### PR DESCRIPTION
**Pull request by Chillee**
Still some stuff to fix up, will finish later.

### CODE REVIEW

File: test/test_decomp.py [hunk 1]
```diff
@@ -265,18 +267,20 @@ def normalize_op_input_output(f, sample, requires_grad=True):
...
-    ("cpu", torch.bfloat16, "addmm"),  # decomposition loses precision
-    ("cpu", torch.bfloat16, "softmax"),  # needs relaxed prec
-    ("cpu", torch.bfloat16, "log_softmax"),  # needs relaxed prec
     # CUBLAS_STATUS_NOT_SUPPORTED when calling
     # `cublasGemmStridedBatchedExFix(handle, opa, opb, (int)m, (int)n, (int)k,
     # (void*)&falpha, a, CUDA_R_16BF, (int)lda, stridea, b, CUDA_R_16BF,
     # (int)ldb, strideb, (void*)&fbeta, c, CUDA_R_16BF, (int)ldc, stridec,
     # (int)num_batches, CUDA_R_32F, CUBLAS_GEMM_DEFAULT_TENSOR_OP)`
     ("cuda", torch.bfloat16, "nn.functional.bilinear"),
+    # randomness
```
Comment (ezyang) [inline test/test_decomp.py:276]: oh, if the rng is the same in the decomp and the kernel we can fix this by just snapshotting rng state, maybe? IDK how practical it is to actually manage this. Feel free to delete the comment below this

File: test/test_decomp.py [hunk 2]
```diff
@@ -474,11 +471,11 @@ def check_decomposed(aten_name):
                 )
 
     def test_torchscriptable(self, device):
-        skip_list = []
+        skip_list = [aten.rsub.Scalar]
```
Comment (ezyang) [inline test/test_decomp.py:474]: how comeeee

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -52,31 +52,46 @@ def decrease_prec(x):
...
+        x = x.unsqueeze(-1)
+    return x
+
 
 @register_decomposition(aten.tanh_backward)
 @cast_for_opmath
-def tanh_backward_decomposition(out_grad: Tensor, y: Tensor):
-    return out_grad * (1 - y * y)
+def tanh_backward(out_grad: Tensor, y: Tensor):
+    return out_grad * (1 - y * y).conj_physical()
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:65]: I wanted to say "why can't we use conj" but the problem is that at the point decompositions run there is no way to apply conjugate views anymore? We should at least make sure calling conj() errors (but it would nice to make this work). cc @bdhirsh since functionalization kind of has a similar problem.

File: torch/_decomp/decompositions.py [hunk 4]
```diff
@@ -101,19 +116,31 @@ def elu_backward_decomposition(
...
         (self > -3.0) & (self < 3.0),
         grad_output * (1.0 / 6.0),
         grad_output.new_zeros(()),
     )
 
 
+@register_decomposition(aten.hardtanh)
+@cast_for_opmath
+def hardtanh(self: Tensor, min_val: float = -1, max_val: float = 1) -> Tensor:
+    return torch.clamp(self, min_val, max_val)
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:138]: is this really what hardtanh is lol

Comment (ezyang) [reply torch/_decomp/decompositions.py:138]: it is lol WTF

File: torch/_decomp/decompositions.py [hunk 5]
```diff
@@ -268,6 +307,49 @@ def log_sigmoid_backward(grad_output: Tensor, self: Tensor, buffer: Tensor) -> T
...
+        return loss
+
+
+def to_real_dtype(dtype: torch.dtype):
+    if dtype == torch.complex32:
+        return torch.float16
+    elif dtype == torch.complex64:
+        return torch.float32
+    elif dtype == torch.complex128:
+        return torch.float64
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:325]: this should be in the real library smh

File: torch/_decomp/decompositions.py [hunk 6]
```diff
@@ -268,6 +307,49 @@ def log_sigmoid_backward(grad_output: Tensor, self: Tensor, buffer: Tensor) -> T
...
+
+
+@register_decomposition(aten.l1_loss)
+def l1_loss(self: Tensor, target: Tensor, reduction: int = Reduction.MEAN.value) -> Tensor:
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:329]: it shouldn't be necessary to specify default arguments explicitly here; the decomp runner should fill them in automatically from schema

Comment (Chillee) [reply torch/_decomp/decompositions.py:329]: Is there any reason to not do it? I could imagine in other situations (such as directly using it), the user might not have the defaults.

File: test/test_decomp.py (hunk 2 above)
Comment (Chillee) [reply test/test_decomp.py:474]: Since our decomposition is relying on implicit scalar => tensor promotion to work haha. But Torchscript has not overload for sub(scalar, tensor)

File: torch/_decomp/decompositions.py [hunk 7]
```diff
@@ -268,6 +307,49 @@ def log_sigmoid_backward(grad_output: Tensor, self: Tensor, buffer: Tensor) -> T
...
+    loss = (self - target).abs()
+    # PyTorch semantics result in the output of l1_loss having the corresponding
+    # real dtype to self.  This may not happen without explicit casting if say
+    # self: complex64 and target: float64, which results in loss: float64
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:333]: that's weird though, abs() is supposed to take complex to real so shouldn't this happen automatically

File: torch/_decomp/decompositions.py [hunk 8]
```diff
@@ -291,6 +382,59 @@ def huber_loss_backward(
     )
 
 
+@register_decomposition(aten.nll_loss_backward)
+def nll_loss_backward(grad_output: Tensor, self: Tensor, target: Tensor, weight: Optional[Tensor], reduction: int, ignore_index: int, total_weight: Tensor) -> Tensor:
+    assert 0 <= self.dim() <= 2, "input tensor should be 1D or 2D"
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:387]: we probably shouldn't actually use asserts for error checking, as they raise AssertionError which won't match the real deal (RuntimeError)

File: torch/_decomp/decompositions.py [hunk 9]
```diff
@@ -309,6 +453,18 @@ def binary_cross_entropy_backward(
     return result
 
 
+@register_decomposition(aten._euclidean_dist)
+def _euclidean_dist(x1: Tensor, x2: Tensor) -> Tensor:
+    x1_norm = x1.pow(2).sum(-1, True)
+    x1_pad = torch.ones_like(x1_norm, memory_format=torch.contiguous_format)
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:459]: This makes me wonder if we should be checking that the reference implementation matches memory layout of the real one :/

File: torch/_decomp/decompositions.py [hunk 10]
```diff
@@ -765,7 +1010,7 @@ def cudnn_batch_norm(
     )
     # Cudnn return running mean and variance when training is True
     if training:
-        return (a, b, c, input.new_zeros((1,)))
+        return (a, b, c, input.new_zeros((0,), dtype=torch.uint8))
```
Comment (ezyang) [inline torch/_decomp/decompositions.py:1013]: that's whack

**Review by ezyang — APPROVED** [review]

File: torch/_decomp/decompositions.py (hunk 6 above)
Comment (ezyang) [reply torch/_decomp/decompositions.py:329]: I suppose the hazard is mostly violation of DRY. I guess if our defaults never change it is relatively harmless.

File: torch/_decomp/decompositions.py (hunk 10 above)
Comment (Chillee) [reply torch/_decomp/decompositions.py:1013]: Tragically, we’re not very consistent about what we return when a tensor is not needed.

Also see aten.native_batch_norm where I need to return something different for saved_mean/saved_invstd depending on whether it’s cuda or cpu.

File: torch/_decomp/decompositions.py (hunk 7 above)
Comment (Chillee) [reply torch/_decomp/decompositions.py:333]: Oh, it does, the problem is that self can be complex64, and then target is float64, so after we do (loss - target) we end up with a complex128 value. Then, calling abs on that gives us a float64, which is not the corresponding real type for complex64.

File: torch/_decomp/decompositions.py (hunk 8 above)
Comment (Chillee) [reply torch/_decomp/decompositions.py:387]: Should we have some kind of “assert” operator? Might be difficult to model in a functional IR.

Comment (ezyang) [reply torch/_decomp/decompositions.py:387]: We need an assert operator, not for this line (since we are expecting to always be rank specialized), but if we do anything with a dynamic shape we can't actually do the assert, we have to encode it into the IR. Functional IRs that don't care about validity can always drop it. Also cc @mruberry who has some thoughts on this.

Comment (mruberry) [reply torch/_decomp/decompositions.py:387]: Having a special assertion operator, even if it only appears in traces, is an interesting idea. I don't know if we need one but it could be very helpful.

Comment (Chillee) [reply torch/_decomp/decompositions.py:387]: I think having one is unavoidable. Asserts on rank may be evaluatable at trace/compile time, we *might* be able to get away with lifting dynamic shapes asserts up, but there other asserts like “assert x.min() > 0” in operators.

If we want to accurately model preconditions like that (and I’d argue we do!), then we need an assertion operator.

Comment (mruberry) [reply torch/_decomp/decompositions.py:387]: We can understand the trace validity of expressions like `assert x.min() > 0` without modeling the assertion, though

Comment (Chillee) [reply torch/_decomp/decompositions.py:387]: What does “understanding the trace validity mean”? Like, figuring out if you need to respecialize? (If so, I agree).

My point was that if we want to model the error checking a function has, we need to model the assert. It’s reasonable to say for now that “we’re just not gonna model the assert”, but eventually, we probably want to.

Comment (mruberry) [reply torch/_decomp/decompositions.py:387]: > What does “understanding the trace validity mean”? Like, figuring out if you need to respecialize? (If so, I agree).

More generally, determining if you don't have a valid trace and so need to retrace, but yes.


> My point was that if we want to model the error checking a function has, we need to model the assert. It’s reasonable to say for now that “we’re just not gonna model the assert”, but eventually, we probably want to.

Idk. With the prims we have meta functions for all their error checking and shape logic. In the future we could even look at separating those two. We'll probably run those checks during trace acquisition.

Comment (Chillee) [reply torch/_decomp/decompositions.py:387]: > We'll probably run those checks during trace acquisition.

How can we run a check like “x.min() > 0” during trace acquisition?

Comment (mruberry) [reply torch/_decomp/decompositions.py:387]: Because we separate the error checking logic from the function's logic, so when we know the check is valid we just call the function without error checking, and if we don't know we run the check

Comment (mruberry) [reply torch/_decomp/decompositions.py:387]: Sorry so in this case we have data-dependent control flow, so we'll almost certainly break the trace at data-dependent control flow. But if this were metadata-dependent control flow then my above statements apply (I think)

Comment (Chillee) [reply torch/_decomp/decompositions.py:387]: > But if this were metadata-dependent control flow then my above statements apply (I think)

Yes I agree with this.

> so we'll almost certainly break the trace at data-dependent control flow.

What I'm saying is that we have an assert operator, we might be able to not break the trace at this kind of data-dependent control flow :) And I think we do have cases like that. At the very least, we might want to capture it into our trace simply to have the option to avoid CUDA syncs.

Comment (Chillee) [reply torch/_decomp/decompositions.py:387]: As for the original discussion, I think I'm gonna leave the way it is now - any other solution is gonna introduce *actual* control flow, which I think is worse than raising an AssertionError instead of a RuntimeError.

File: torch/_decomp/decompositions.py (hunk 6 above)
Comment (Chillee) [reply torch/_decomp/decompositions.py:329]: If we had a (from python) way of calling an operator with the decompositions filled in, I'd be happy to remove them. 

I'd also be happy to remove them if they ever cause us problems in the future lol.

File: torch/_decomp/decompositions.py (hunk 7 above)
Comment (Chillee) [reply torch/_decomp/decompositions.py:336]: By the way, if it makes you feel better, our actual kernel does something similar :P 

https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Loss.cpp#L521
